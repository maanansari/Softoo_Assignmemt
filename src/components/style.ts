import {StyleSheet} from 'react-native';

export const style = StyleSheet.create({
  mainContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    borderRadius: 10,
    backgroundColor: '#0F52BA',
    paddingVertical: 2,
    marginHorizontal: 4,
    paddingHorizontal: 5,
    alignItems: 'center',
  },
  textContainer: {
    flex: 0.75,
    color: '#FFF',
    fontWeight: '500',
    fontSize: 17,
  },
  priceTextStyle: {
    fontSize: 15.5,
    flex: 0.25,
    textAlign: 'right',
  },
  storeImageStyle: {
    width: 400,
    height: 400,
    alignSelf: 'center',
  },
  removeCartContainer: {
    alignSelf: 'center',
    alignItems: 'center',
    flex: 1,
  },
  actionLayoutStyle: {
    flexDirection: 'row',
    alignSelf: 'center',
    alignItems: 'center',
  },
  countTextStyle: {
    fontSize: 16.5,
    marginHorizontal: 8,
  },
  buttonStyle: {
    alignSelf: 'center',
    borderRadius: 5,
    paddingHorizontal: 20,
    paddingVertical: 7,
    marginVertical: 2,
    marginHorizontal: 4,
    backgroundColor: '#0F52BA',
  },
  buttonTextStyle: {
    color: '#FFF',
    fontSize: 17,
  },
});
