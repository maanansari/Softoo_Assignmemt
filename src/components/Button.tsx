import React from 'react';
import {Text, TouchableOpacity} from 'react-native';
import {style} from './style';

type Props = {
  title: string;
  onPress: () => void;
};

const Button: React.FC<Props> = ({title, onPress}) => {
  return (
    <TouchableOpacity testID="btnContainer" onPress={onPress} style={style.buttonStyle}>
      <Text testID="btnText" style={style.buttonTextStyle}>
        {title}
      </Text>
    </TouchableOpacity>
  );
};

export default Button;
