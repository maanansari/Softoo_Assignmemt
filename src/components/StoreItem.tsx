import {Image, Text, View} from 'react-native';
import {useDecrement, useIncrement, useRemove} from '../hooks/useHooks';
import {useReduxDispatch, useReduxSelector} from '../redux/index';
import Button from './Button';
import {style} from './style';

export function StoreItems({products}: any) {
  const {id, name, price, img} = products;
  const dispatch = useReduxDispatch();
  const val = useReduxSelector(sel => sel.cart);
  const quantity = val.find(item => item?.products.id == id)?.products?.quantity || 0;

  const handleIncreaseQuantity = () => {
    dispatch(useIncrement(products));
  };

  const handleDecreaseQuantity = () => {
    dispatch(useDecrement(products));
  };

  const handleRemoveFromCart = () => {
    dispatch(useRemove(products));
  };

  return (
    <View style={{flex: 0.5}}>
      <Image source={{uri: img}} style={style.storeImageStyle} resizeMode="contain" />
      <View style={style.mainContainer}>
        <Text style={style.textContainer} numberOfLines={1}>
          {name}
        </Text>
        <Text style={[style.textContainer, style.priceTextStyle]}>${price}</Text>
      </View>
      {quantity === 0 || quantity === undefined ? (
        <Button title="Add to cart" onPress={handleIncreaseQuantity} />
      ) : (
        <View style={style.removeCartContainer}>
          <View style={style.actionLayoutStyle}>
            <Button title="+" onPress={handleIncreaseQuantity} />
            <Text testID="quantityText" style={style.countTextStyle}>{`${quantity} in cart`}</Text>
            <Button title="-" onPress={handleDecreaseQuantity} />
          </View>
          <Button title="Remove from cart" onPress={handleRemoveFromCart} />
        </View>
      )}
    </View>
  );
}
