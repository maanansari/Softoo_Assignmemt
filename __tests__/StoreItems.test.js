import {configureStore} from '@reduxjs/toolkit';
import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {Provider} from 'react-redux';
import {StoreItems} from '../src/components/StoreItem';
import rootReducer from '../src/redux/rootReducer';

describe('StoreItems', () => {
  let store = configureStore({
    reducer: rootReducer,
  });

  const products = {
    id: 1,
    name: 'Product 1',
    price: 10,
    img: 'https://example.com/product1.jpg',
  };

  let storeComponent = (
    <Provider store={store}>
      <StoreItems products={products} />
    </Provider>
  );

  it('match snapshot', () => {
    const {toJSON} = render(storeComponent);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly and handles actions', async () => {
    const {findByText} = render(storeComponent);
    const addToCartButton = await findByText('Add to cart');
    expect(addToCartButton).toBeDefined();
    fireEvent.press(addToCartButton);
    expect(await findByText('-')).toBeDefined();
    expect(await findByText('+')).toBeDefined();
    expect(await findByText('Remove from cart')).toBeDefined();
  });

  it('renders correctly and handles actions', async () => {
    const {findByTestId} = render(storeComponent);
    const quantityText = await findByTestId('quantityText');
    expect(quantityText.props.children).toContain('1 in cart');
  });
});
