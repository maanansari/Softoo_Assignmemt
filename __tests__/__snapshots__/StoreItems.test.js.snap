// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreItems match snapshot 1`] = `
<View
  style={
    {
      "flex": 0.5,
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "https://example.com/product1.jpg",
      }
    }
    style={
      {
        "alignSelf": "center",
        "height": 400,
        "width": 400,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#0F52BA",
        "borderRadius": 10,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginHorizontal": 4,
        "paddingHorizontal": 5,
        "paddingVertical": 2,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#FFF",
          "flex": 0.75,
          "fontSize": 17,
          "fontWeight": "500",
        }
      }
    >
      Product 1
    </Text>
    <Text
      style={
        [
          {
            "color": "#FFF",
            "flex": 0.75,
            "fontSize": 17,
            "fontWeight": "500",
          },
          {
            "flex": 0.25,
            "fontSize": 15.5,
            "textAlign": "right",
          },
        ]
      }
    >
      $
      10
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "center",
        "backgroundColor": "#0F52BA",
        "borderRadius": 5,
        "marginHorizontal": 4,
        "marginVertical": 2,
        "opacity": 1,
        "paddingHorizontal": 20,
        "paddingVertical": 7,
      }
    }
    testID="btnContainer"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 17,
        }
      }
      testID="btnText"
    >
      Add to cart
    </Text>
  </View>
</View>
`;
